From 0ca9a9d6be9949b744cedd485ea2c52f8a39d288 Mon Sep 17 00:00:00 2001
From: TrueNAS Patch Series <patchseries@vbpub.invalid>
Date: Fri, 30 Oct 2026 00:00:00 +0000
Subject: [PATCH] Mount boot-pool user data during boot

Mount the managed user-data root during the pool import boot sequence.

Change-Id: v28-boot-mount
Capability: boot-pool-user-data
Role: enable
Risk: medium
Requires: v28-storage
Forwarded: not-needed
---
--- a/middlewared/plugins/pool_/import_pool.py
+++ b/middlewared/plugins/pool_/import_pool.py
@@ -23,6 +23,7 @@
 from middlewared.utils.filesystem import attrs as fs_attrs
 from middlewared.utils.zfs import query_imported_fast_impl
 
+from .boot_pool_mount import mount_boot_pool_user_data
 from .utils import ZPOOL_CACHE_FILE
 
 
@@ -659,6 +660,13 @@
 
         # TODO: we need to fix this. There is 0 reason to do all this stuff
         # and block the entire boot-up process.
+        # boot-pool is not registered in storage.volume, so mount its opted-in
+        # user-data namespace explicitly after normal data-pool imports.
+        try:
+            mount_boot_pool_user_data(self)
+        except Exception:
+            self.logger.error('Unexpected error mounting boot-pool user-data datasets', exc_info=True)
+
         self.logger.debug('Calling pool.post_import')
         self.middleware.call_hook_sync('pool.post_import', None)
         self.logger.debug('Finished calling pool.post_import')
